### PR TITLE
download the latest Firefox coverage build

### LINKED ISSUE #5: Write script to download the latest Firefox coverage build

### PR DESCRIPTION
**Pull request by rhcu**
Fixes #5.
 
The latest coverage build is downloaded by taskId retrieved from Index API of Taskcluster; extracted to the directory script was run, and the downloaded file is deleted after that. 

I have used ideas from your `firefox-code-coverage` repository from this file: https://github.com/marco-c/firefox-code-coverage/blob/5c8bc65ab17aa101183fdae5ed56fddf210a3620/codecoverage.py

### CODE REVIEW

**Comment by marco-c** [comment]
Hey @rhcu, that script is made to be self-contained to avoid third-party dependencies (so users of that script don't need to install anything other than Python). In this case, we don't care about that as we will definitely require many libraries to be installed.

Can you try to rewrite the script to use the taskcluster-client library?

**Comment by rhcu** [comment]
hi @marco-c, thanks for a comment. I can try to rewrite it. Basically, functionality will be the same as in this one?

**Comment by marco-c** [comment]
Yes, same functionality but using the taskcluster client library.

**Comment by marco-c** [comment]
Thanks, this is nicer :)

File: latest_cov_build.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import tarfile
```
Comment (marco-c) [inline latest_cov_build.py]: Make sure there are no errors when running `flake8` on this file (and fix any error that will come up)

File: latest_cov_build.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    from urllib.request import urlopen, urlretrieve
+except ImportError:
+    from urllib import urlencode, urlretrieve
+    from urllib2 import urlopen
+
+
+index = taskcluster.Index()
+queue = taskcluster.Queue()
+route = "gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt"
+taskId = index.findTask(route)['taskId']
```
Comment (marco-c) [inline latest_cov_build.py]: Since it's only used here, put the value here directly and remove the route variable.
Use single quotes everywhere for strings instead of double quotes.

File: latest_cov_build.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+try:
+    from urllib.parse import urlencode
+    from urllib.request import urlopen, urlretrieve
+except ImportError:
+    from urllib import urlencode, urlretrieve
+    from urllib2 import urlopen
+
+
+index = taskcluster.Index()
+queue = taskcluster.Queue()
```
Comment (marco-c) [inline latest_cov_build.py]: Define the variable where you are using it

Comment (marco-c) [reply latest_cov_build.py]: This way the code gets easier to read

File: latest_cov_build.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+route = "gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt"
+taskId = index.findTask(route)['taskId']
+artifactName = 'public/build/target.tar.bz2'
+name = os.path.join('ccov-artifacts', taskId +
+                    'artifacts.public.build.target.tar.bz2')
+url = queue.buildUrl('getLatestArtifact', taskId, artifactName)
+print(url)
```
Comment (marco-c) [inline latest_cov_build.py]: Remove the logging statements

File: latest_cov_build.py (hunk 1 above)
Comment (marco-c) [reply latest_cov_build.py]: Also, please divide the file in three sections: 1) Retrieve task from Index; 2) Retrieve artifact; 3) Unpack artifact

To do this either define functions and call them or do something like this:
```
# Retrieving task from index
CODE HERE

# Retrieving artifact from Queue
CODE HERE

# Extracting artifact
CODE HERE
```

File: latest_cov_build.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+urlretrieve(url, name)
+tar = tarfile.open(name, 'r:bz2')
```
Comment (marco-c) [inline latest_cov_build.py]: Please use the context manager here, so you don't need to close tar explicitly

File: latest_cov_build.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+
+index = taskcluster.Index()
+queue = taskcluster.Queue()
+route = "gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt"
+taskId = index.findTask(route)['taskId']
+artifactName = 'public/build/target.tar.bz2'
+name = os.path.join('ccov-artifacts', taskId +
+                    'artifacts.public.build.target.tar.bz2')
+url = queue.buildUrl('getLatestArtifact', taskId, artifactName)
```
Comment (marco-c) [inline latest_cov_build.py]: Have you tried `queue.getLatestArtifact`?

Comment (marco-c) [reply latest_cov_build.py]: There might even be a way to retrieve the artifact directly without getting the task first: https://github.com/taskcluster/taskcluster-client.py#get-artifact-from-indexed-task.

Comment (rhcu) [reply latest_cov_build.py]: Thanks for previous comments. :) 
No, I haven't tried because as I have understood from Taskcluster API it was written: `API Clients, this method will redirect you to the artifact, if it is stored externally. Either way, the response may not be JSON. So API client users might want to generate a signed URL for this end-point and use that URL with a normal HTTP client.` - so, I've decided that it I will anyway need to build URL.
But I can try to redo it with this method if it is better. :)

Comment (marco-c) [reply latest_cov_build.py]: Yes, it would be better as it is more clear

File: README.md [hunk 7]
```diff
@@ -1,2 +1,5 @@
 # coverage-crawler
+
+[![Build Status](https://travis-ci.org/marco-c/coverage-crawler.svg?branch=master)](https://travis-ci.org/marco-c/coverage-crawler)
```
Comment (marco-c) [inline README.md]: This should be part of a separate PR, we should try to make each PR self-contained

File: latest_cov_build.py (hunk 6 above)
Comment (rhcu) [reply latest_cov_build.py]: I wanted to ask here:
 1) I have tried this method https://github.com/taskcluster/taskcluster-client.py#get-artifact-from-indexed-task, but it hasn't worked. When I've tried to use the link constructed by pattern described in the documentation of this method, it returns Error. Here it is: https://index.taskcluster.net/task/gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt/artifacts/public/target.tar.bz2
2) queue.getLatestArtifact() method is void, and I have seen in documentation (honestly, for Javascript) that they use queue.buildUrl('getLatestArtifact', taskId, artifactName) method for getting url.

File: README.md (hunk 7 above)
Comment (rhcu) [reply README.md]: Should I change this now or prevent a mess in the next time?

Comment (marco-c) [reply README.md]: Change this now and move it to a separate PR

File: latest_cov_build.py (hunk 6 above)
Comment (marco-c) [reply latest_cov_build.py]: 1 is strange, the URL should be `https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt/artifacts/public/build/target.tar.bz2`. Have you tried `index.findArtifactFromTask('gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt', 'public/build/target.tar.bz2')`? Or you can do `urlretrieve('https://index.taskcluster.net/v1/task/gecko.v2.mozilla-central.latest.firefox.linux64-ccov-opt/artifacts/public/build/target.tar.bz2')` directly
2 yes they use that behind the scenes, but it could change in the future so it's better to use their interface rather than reimplementing it.

File: README.md (hunk 7 above)
Comment (rhcu) [reply README.md]: @marco-c I am sorry for such messy PR. I've changed the code to use just urlretrieve as you suggested, the taskId is used for unique identification of coverage builds as tar.bz2 archive.

File: README.md [hunk 8]
```diff
@@ -1,4 +1,5 @@
 # coverage-crawler
+
```
Comment (marco-c) [inline README.md]: Can you remove this unrelated change too?

File: latest_cov_build.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    from urllib.request import urlretrieve
+except ImportError:
+    from urllib import urlretrieve
+
+
+# Retrieving task from index
+index = taskcluster.Index()
+taskId = index.findTask('gecko.v2.mozilla-central.' +
+                        'latest.firefox.linux64-ccov-opt')['taskId']
+# taskId uniquelly identifies artifact
```
Comment (marco-c) [inline latest_cov_build.py]: There's no need for that, we always want the latest build so we don't need to store the old ones too

File: latest_cov_build.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+name = os.path.join('ccov-artifacts', taskId +
+                    'artifacts.public.build.target.tar.bz2')
+# Retrieving artifact
+urlretrieve('https://index.taskcluster.net/v1/task/gecko.v2.' +
+            'mozilla-central.latest.firefox.linux64-ccov-opt/' +
+            'artifacts/public/build/target.tar.bz2', name)
+# Extracting artifact
```
Comment (marco-c) [inline latest_cov_build.py]: Put a newline before the comments, so the code is easier to read
